csrc/multidevice/communication.cpp

```diff
@@ -138,6 +281,63 @@ c10::intrusive_ptr<c10d::Work> Broadcast::post(
 }
 
 Gather::Gather(CommParams params) : Communication(params, "gather") {
+  validateParams();
+}
+
+Gather::Gather(
+    TensorView* input_tv,
+    TensorView* output_tv,
+    at::Tensor input,
+    at::Tensor output,
+    DeviceIdxType my_device_index,
+    DeviceIdxType root)
+    : Communication("gather") {
+  params_.root = root;
+  DeviceMesh mesh = input_tv->getDeviceMesh();
+  params_.team = mesh.vector();
+  bool is_root_in_mesh = mesh.has(root);
+  if (!is_root_in_mesh) {
+    params_.team.push_back(root);
+  }
+
+  if (my_device_index == root &&
+      requiresRelayoutOutputTensor(input_tv, output_tv)) {
+    output_ = output;
+    output = allocateOutputTensor(input_tv, output);
```

Comment:
It is probably ok for now but ideally all the allocations should be gathered into the allocator. It can be misleading that instantiating the class implicitely triggers an allocation under the hood. We should keep in mind that, because allocation kernel calls are expensive, we must not instantiate collective at fusion execution

Reply:
It is even worse...the contiguous() calls on the input tensors will also trigger a copy if the tensor is not contiguous.
I agree they should be gathered by the allocator.